jerry-debugger/src/lib/protocol-handler.ts

```diff
@@ -0,0 +1,251 @@
+
+import * as SP from './jrs-protocol-constants';
+import { ParsedFunction } from './breakpoint';
+import { ByteConfig, cesu8ToString, concatUint8Arrays, decodeMessage } from './utils';
+
+// expected JerryScript debugger protocol version
+const JERRY_DEBUGGER_VERSION = 1;
+
+export interface ParserStackFrame {
+  isFunc: boolean;
+  line: number;
+  column: number;
+  name: string;
+  source: string;
+  sourceName?: string;
+  lines: Array<number>;
+  offsets: Array<number>;
+  byteCodeCP?: number;
+  firstBreakpointLine?: number;
+  firstBreakpointOffset?: number;
+}
+
+export interface JerryDebugProtocolDelegate {
+  onError?(message: string): void;
+  onScriptParsed?(message: JerryMessageScriptParsed): void;
+}
+
+export interface JerryMessageScriptParsed {
+  scriptId: number;
+  name: string;
+  endLine: number;
+  endColumn: number;
+  hasSourceURL: boolean;
+}
+
+interface ProtocolFunctionMap {
+  [type: number]: (data: Uint8Array) => void;
+}
+
+interface FunctionMap {
+  [cp: string]: ParsedFunction;
+}
+
+// abstracts away the details of the protocol
+export class JerryDebugProtocolHandler {
+  public delegate: JerryDebugProtocolDelegate;
+
+  // debugger configuration
+  private maxMessageSize: number = 0;
+  private byteConfig: ByteConfig;
+  private version: number = 0;
+  private functionMap: ProtocolFunctionMap;
+
+  private stack: Array<ParserStackFrame>;
+  private source: string = '';
+  private sourceData?: Uint8Array;
+  private sourceName?: string;
+  private sourceNameData?: Uint8Array;
+  private functionName?: Uint8Array;
+  private functions: FunctionMap = {};
+  private lastScriptID: number = 0;
+
+  constructor(delegate: JerryDebugProtocolDelegate) {
+    this.delegate = delegate;
+
+    this.byteConfig = {
+      cpointerSize: 0,
+      littleEndian: true,
+    };
+
+    this.functionMap = {
+      [SP.JERRY_DEBUGGER_CONFIGURATION]: this.onConfiguration,
+      [SP.JERRY_DEBUGGER_BYTE_CODE_CP]: this.onByteCodeCP,
+      [SP.JERRY_DEBUGGER_PARSE_FUNCTION]: this.onParseFunction,
+      [SP.JERRY_DEBUGGER_BREAKPOINT_LIST]: this.onBreakpointList,
+      [SP.JERRY_DEBUGGER_BREAKPOINT_OFFSET_LIST]: this.onBreakpointList,
+      [SP.JERRY_DEBUGGER_SOURCE_CODE]: this.onSourceCode,
+      [SP.JERRY_DEBUGGER_SOURCE_CODE_END]: this.onSourceCode,
+      [SP.JERRY_DEBUGGER_SOURCE_CODE_NAME]: this.onSourceCodeName,
+      [SP.JERRY_DEBUGGER_SOURCE_CODE_NAME_END]: this.onSourceCodeName,
+    };
+
+    this.stack = [{
+      isFunc: false,
+      line: 1,
+      column: 1,
+      name: '',
+      source: '',
+      lines: [],
+      offsets: [],
+    }];
+  }
+
+  // for the temporary mollification of lint
+  unused() {
+    this.maxMessageSize;
+  }
+
+  decodeMessage(format: string, message: Uint8Array, offset: number) {
+    return decodeMessage(this.byteConfig, format, message, offset);
+  }
+
+  onConfiguration(data: Uint8Array) {
+    console.log('[Configuration]');
+    if (data.length !== 5) {
```

Comment:
Maybe do `<` instead?
In theory, if the configuration message changed, the version should have been bumped.
I think it's safe to assume the message won't be *shorter* in newer versions.
I think it would be more friendly/useful to hit the "incorrect target debugger version" error than to get "configuration message wrong size".